aplus/celery.py

```diff
@@ -32,3 +40,25 @@ def enroll():
         i.enroll_from_sis()
         if settings.SIS_ENROLL_DELAY:
             sleep(settings.SIS_ENROLL_DELAY)
+
+@app.task
+def retry_submissions():
+    from exercise.submission_models import PendingSubmission, PendingSubmissionManager
+
+    # Recovery state: only send one grading request to probe the state of grader
+    # We pick the one with most attempts, so that total_retries comes down more quickly
+    # when things are back to normal, and we can return to normal state
+    if not PendingSubmissionManager.is_grader_stable():
+        pending = PendingSubmission.objects.order_by('-num_retries').first()
+        logging.info(f"Recovery state: retrying expired submission {pending.submission}")
+        pending.submission.exercise.grade(pending.submission)
+        return
+
+    # Stable state: retry all expired submissions
+    expiry_time = datetime.datetime.now(datetime.timezone.utc) - relativedelta(seconds=settings.SUBMISSION_EXPIRY_TIMEOUT)
+    expired = PendingSubmission.objects.filter(submission_time__lt=expiry_time)
+
+    for pending in expired:
+        logger.info(f"Retrying expired submission {pending.submission}")
+        pending.submission.exercise.grade(pending.submission)
+        sleep(0.5)  # Delay 500 ms to avoid choking grader
```

Comment:
```suggestion
        if pending.submission.exercise.can_regrade:
            logger.info(f"Retrying expired submission {pending.submission}")
            pending.submission.exercise.grade(pending.submission)
            sleep(0.5)  # Delay 500 ms to avoid choking grader
```
LTI exercises can not be regraded since the submission is not stored in A+; the student sends it directly to the LTI Tool. Though, LTI exercise submissions should never have any PendingSubmission. However, A+ uses the LTIExercise model for Rubyric exercises, but since there is a field `aplus_get_and_post`, then Rubyric exercises still behave almost like normal exercises in grading. But regrading manually graded exercises in Rubyric would not make sense. We don't want to automatically send multiple copies of the same submission to Rubyric until some teacher has manually graded it.
https://github.com/apluslms/a-plus/blob/25e6a22f43fdb283e9a4d6330dc70b68f179630b/exercise/exercise_models.py#L1013-L1016

Reply:
Another example of manually graded exercises: Jutut feedback questionnaires. In the A+ database, they are BaseExercises, no LTI there. We can't regrade Jutut questionnaires since we don't know if the teacher intends to manually grade them at all.